pkg/container-hook/bpf/execruntime.bpf.c

```diff
@@ -0,0 +1,221 @@
+	if (task != current_task)
+		return 0;
+
+	if (tracer_fanotify_fd == ctx->fd) {
+		// found it!
+		tracer_group = (u64) BPF_CORE_READ(file, private_data);
+		return 0;
+	}
+
+	return 0;
+}
+
+SEC("kprobe/fsnotify_remove_first_event")
+int BPF_KPROBE(ig_fa_pick, struct fsnotify_group *group)
+{
+	u64 pid_tgid;
+	u64 dummy = 0;
+
+	if (tracer_group != (u64)group)
+		return 0;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+
+	// Keep context for kretprobe/fsnotify_remove_first_event
+	bpf_map_update_elem(&ig_fa_pick_ctx, &pid_tgid, &dummy, 0);
+
+	return 0;
+}
+
+SEC("kretprobe/fsnotify_remove_first_event")
+int BPF_KRETPROBE(ig_fa_pick_x, struct fanotify_event *ret)
+{
+	struct record *record;
+	u32 pid;
+	u64 pid_tgid;
+	u64 *exists;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+
+	exists = bpf_map_lookup_elem(&ig_fa_pick_ctx, &pid_tgid);
+	if (!exists)
+		return 0;
+
+	pid = BPF_CORE_READ(ret, pid, numbers[0].nr);
+
+	record = bpf_map_lookup_elem(&exec_args, &pid);
+	if (!record) {
+		// no record found but we need to push an empty record in the queue to
+		// ensure userspace understands that there is no record for this event
+		goto add_empty_record;
+	}
+
+	bpf_map_push_elem(&ig_fa_records, record, 0);
+
+	bpf_map_delete_elem(&ig_fa_pick_ctx, &pid_tgid);
+	return 0;
+
+add_empty_record:
+	bpf_map_push_elem(&ig_fa_records, &empty_record, 0);
+	bpf_map_delete_elem(&ig_fa_pick_ctx, &pid_tgid);
+	return 0;
+}
+
+SEC("tracepoint/syscalls/sys_enter_execve")
+int ig_execve_e(struct trace_event_raw_sys_enter* ctx)
+{
+	u64 pid_tgid;
+	u32 tgid;
+	struct record *record;
+	struct task_struct *task;
+	uid_t uid = (u32)bpf_get_current_uid_gid();
+
+	unsigned int ret;
+	const char **args = (const char **)(ctx->args[1]);
+	const char *argp;
+	int i;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	tgid = pid_tgid >> 32;
+
+	// Add new entry but not from the stack due to size limitations
+	if (bpf_map_update_elem(&exec_args, &tgid, &empty_record, 0))
+		return 0;
+	record = bpf_map_lookup_elem(&exec_args, &tgid);
+	if (!record)
+		return 0;
+
+	task = (struct task_struct*)bpf_get_current_task();
+
+	record->timestamp = bpf_ktime_get_boot_ns();
+	bpf_get_current_comm(&record->caller_comm, sizeof(record->caller_comm));
+	record->pid = tgid;
+	record->uid = uid;
+	record->ppid = (pid_t)BPF_CORE_READ(task, real_parent, tgid);
+	record->args_count = 0;
+	record->args_size = 0;
+
+	ret = bpf_probe_read_user_str(record->args, ARGSIZE, (const char*)ctx->args[0]);
+	if (ret <= ARGSIZE) {
+		record->args_size += ret;
+	} else {
+		// write an empty string
+		record->args[0] = '\0';
+		record->args_size++;
+	}
+
+	record->args_count++;
+	#pragma unroll
+	for (i = 1; i < TOTAL_MAX_ARGS && i < max_args; i++) {
+		bpf_probe_read_user(&argp, sizeof(argp), &args[i]);
+		if (!argp)
+			return 0;
+
+		if (record->args_size > LAST_ARG)
+			return 0;
+
+		ret = bpf_probe_read_user_str(&record->args[record->args_size], ARGSIZE, argp);
+		if (ret > ARGSIZE)
```

Comment:
I think it was working because the return value of `bpf_probe_read_user_str()` is casted into an `unsigned int` so when it returns an error, it is actually a big number, so `ret > ARGSIZE` is true.

However, this code pattern is error-prone, so I will make it more explicit.

Reply:
Filed https://github.com/inspektor-gadget/inspektor-gadget/pull/1763 for the part in trace exec.